Egor Makhlaev

egor-makhlaev opened the pull request:
Python Final Task

- [x] First iteration

1. - [x]   -h, --help     show this help message and exit
2. - [x]   --version      Print version info
3. - [x]   --json         Print result as JSON in stdout
4. - [x]   --verbose      Outputs verbose status messages
5. - [x]   --limit LIMIT  Limit news topics if this parameter provided

- [x] Second iteration
- [x] Third iteration
- [x] Fourth iteration
- [ ] Fifth iteration
- [ ] Sixth iteration

egor-makhlaev committed:
Create version 1.0

egor-makhlaev committed:
Structure project and add tests

egor-makhlaev commented:
Implemented iterations 1-2

egor-makhlaev committed:
Fix style issues in test_main.py and test_print_news.py

egor-makhlaev committed:
Fix typo in test_print_news.py

egor-makhlaev committed:
Fix import error

Shurke commented on rss_reader/rss_reader/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from .rss_reader import main
+from .rss_reader import logger
+from .rss_reader import VERSION
+from .rss_reader import print_news
+from .rss_reader import get_response
+from .rss_reader import process_response
```
Я бы рекомендовал все таки использовать абсолютный импорт, а не явный относительный.
Это позволит в будущем избежать путаницы и иных проблем.

egor-makhlaev replied:
Принял к сведению.

egor-makhlaev replied:
> Я бы рекомендовал все таки использовать абсолютный импорт, а не явный относительный.
> Это позволит в будущем избежать путаницы и иных проблем.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,108 @@
+
+    try:
+        logger.info(f"Try to get response from: {url}")
+
```
Можете пояснить необходимость пустой строки здесь?

egor-makhlaev replied:
Необходимости нет, изначально не досмотрел.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,108 @@
+
+
+def process_response(response, limit):
+    """Returns dictionary with channel info and items. Items number is determined by provided limit argument."""
```
Хорошей практикой является указывать, что за параметры у функции.

```python
def process_response(response, limit):
    """Returns dictionary with channel info and items. Items number is determined by provided limit argument.

    Args:
      response: описание того, что это и какой тип данных.
      limit: описание того, что это и какой тип данных.

    Returns:
      что возвращает функция и какой тип данных у ответа.
    """
```

Это касается и других функций.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,108 @@
+
+
+def get_response(url):
+    """Returns HTTPResponse from server using provided url."""
+    global logger
+
+    try:
+        logger.info(f"Try to get response from: {url}")
+
+        response = urlopen(url)
+
+    except Exception as e:
+        logger.error("Couldn't get good response")
+        sys.exit()
```
Я бы предложил несколько расширить данный фрагмент.
В response содержит такое свойство, как status. И на сновании данного статуса, можно предусмотреть для пользователя более информативное сообщение об ошибке.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,108 @@
+
+def get_response(url):
+    """Returns HTTPResponse from server using provided url."""
+    global logger
```
В этом нет неообходимотси.
Вспомните про LEGB.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,108 @@
+    global logger
+
+    try:
+        logger.info(f"Try to parse response")
```
Обычно такие вещи пишут, как описание того, что сейчас произойдет.
```python
logger.info(f"Parse response")
```

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,108 @@
+    channel_title = root.findtext("channel/title")
+    channel_items = []
+
+    for i, item in enumerate(islice(root.iterfind("channel/item"), 0, max(0, limit) if limit is not None else limit)):
```
i - это плохое название для переменной.
Как собственно и item.

egor-makhlaev replied:
Хотелось написать одной строкой без переносов, поэтому такое бедное наименование переменных. Будет исправлено.

Shurke commented on rss_reader/tests/test_exceptions.py:
```diff
@@ -0,0 +1,36 @@
+import os
+import unittest
+from urllib.request import *
```
Импорт через * является плохой практикой и стоит избегать такого.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/tests/test_main.py:
```diff
@@ -0,0 +1,35 @@
+        self.captured_output = io.StringIO()
+        sys.stdout = self.captured_output
+
+    def test_url_and_version_arguments(self):
+        """Tests that app prints its version and stops if URL and --version arguments are specified."""
+        with self.assertRaises(SystemExit):
+            main([None, "https://news.yahoo.com/rss/", "--version"])
+
+        self.assertEqual(self.captured_output.getvalue(), '"Version 2.0"\n')
```
По факту, тут содержится два различных тестовых случая.
Пожалуйста, разделите данный тест на два раздельных:

1. первый тестирует url
2. второй тестирует версию

И стоит добавить тесты также для других возможных аргументов.

egor-makhlaev replied:
Этим тестом я тестировал то, что при передаче url вместе с --version только печатается версия и работа на этом прекращается. Тест на версию есть, на url добавлю. Стоит ли оставить этот тест?

Shurke replied:
В таком случае да, тест вполне валидный.

Shurke commented on rss_reader/tests/test_process_response.py:
```diff
@@ -0,0 +1,44 @@
+        """Opens connection with example.xml to create fake response for process_response."""
+        self.fake_response = urlopen("file:" + pathname2url(os.path.abspath("data/goodsample.xml")))
+
+    def test_none_limit(self):
+        """Tests process_response from rss_reader with limit set to None."""
+        self.assertEqual(len(process_response(self.fake_response, None)["Items"]), 9, "Wrong output size")
+
+    def test_negative_limit(self):
+        """Tests process_response from rss_reader with limit set to -1."""
+        self.assertEqual(len(process_response(self.fake_response, -1)["Items"]), 0, "Wrong output size")
+
+    def test_zero_limit(self):
+        """Tests process_response from rss_reader with limit set to 0."""
+        self.assertEqual(len(process_response(self.fake_response, 0)["Items"]), 0, "Wrong output size")
+
+    def test_adequate_positive_limit_1(self):
+        """Tests process_response from rss_reader with limit set to 1."""
+        self.assertEqual(len(process_response(self.fake_response, 1)["Items"]), 1, "Wrong output size")
+
+    def test_adequate_positive_limit_9(self):
+        """Tests process_response from rss_reader with limit set to 9."""
+        self.assertEqual(len(process_response(self.fake_response, 9)["Items"]), 9, "Wrong output size")
+
+    def test_inadequate_positive_limit(self):
+        """Tests process_response from rss_reader with limit set to 10."""
+        self.assertEqual(len(process_response(self.fake_response, 10)["Items"]), 9, "Wrong output size")
```
Данные тесты обладают одинковым телом и различаются только набором входных данных.
Советую посмотреть в сторону ddt, чтобы параметризировать тест и избежать дублирования кода.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/tests/test_process_response.py:
```diff
@@ -0,0 +1,44 @@
+import os
+import unittest
+from urllib.request import *
```
Вместо имопорта всех имен из модуля, лучше импортировать ровн то, что вы будете использовать.
```python
from urllib.request import urlopen
```

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,108 @@
+def main(argv=sys.argv):
+    parser = argparse.ArgumentParser(description="Pure Python command-line RSS reader.")
+    parser.add_argument("source", type=str, help="RSS URL")
+    parser.add_argument("--version", action="version", version=f'"Version {VERSION}"', help="Print version info")
+    parser.add_argument("--json", action="store_true", help="Print result as JSON in stdout")
+    parser.add_argument("--verbose", action="store_true", help="Outputs verbose status messages")
+    parser.add_argument("--limit", type=int, help="Limit news topics if this parameter provided")
```
Не хватает следующего аргумента:
-h, --help     show this help message and exit

egor-makhlaev replied:
Аргументы -h, --help с их описанием идут по дефолту в объекте ArgumentParser. Есть ли смысл прописывать их в коде?

Shurke replied:
Нет, не нужно.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,108 @@
+
+    print(f"\nFeed: {channel['Title']}")
+
+    for item in channel["Items"]:
+        print("")
+        for key, value in item.items():
+            print(f"{key}: {value}")
+
```
Абсолютно непонятный нейминг.
Исправьте на что-то более вразумительное.

egor-makhlaev replied:
Принял к сведению.

Shurke reviewed (commented):
К замечания изложенным ниже хочу обратить внимание на commit messages.
На данном этапе они не несут никакой информативности о том, что было сделанно в коммите.
Об этом есть пункт в общих требованиях.
Commit messages should provide correct and helpful information about changes in commit. Messages like Fix bug, Tried to make workable, Temp commit and Finally works are prohibited.

egor-makhlaev committed:
Code refactoring

Shurke commented:
Исправьте конфликт для корректной работы CI.

egor-makhlaev committed:
Merge branch 'main' into final-task-implementation

egor-makhlaev committed:
Change nosetests argument

egor-makhlaev committed:
Merge branch 'final-task-implementation' of https://github.com/egor-makhlaev/PythonFinalTask into final-task-implementation

egor-makhlaev committed:
Fix style issues

egor-makhlaev committed:
Fix import error

egor-makhlaev committed:
Add --cover-package argument for nosetests

egor-makhlaev committed:
1. Add exceptions description to get_response function from rss_reader.py
2. Add dynamically generated version string to tests from test_main.py

egor-makhlaev committed:
Fix style issues in test_exceptions.py

egor-makhlaev committed:
Add ddt to test_process_response.py

egor-makhlaev committed:
Update requirements.txt

egor-makhlaev committed:
1. Rename function process_response to parse_response in rss_reader.py
   as now it's only parsing response
2. Add function set_limit to rss_reader.py that sets limit of output news
3. Create local_storage.py with LocalStorage class
4. Add --date argument to parser and corresponding logic for its processing
5. Add news.json and delete goodsample.xml as test logic will be changed

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -49,32 +67,49 @@ def process_response(response, limit):
+
+            Parameters:
+                    {"Title": (str), "Items": (list)}: Dictionary with channel title and items
+            Returns:
+                    (None)
+    """
```
Если функция ничего не возвращает, то не обязательно для нее писать Returns.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/tests/test_process_response.py:
```diff
@@ -1,39 +1,27 @@
+from urllib.request import urlopen
+from urllib.request import pathname2url
+
+from ddt import ddt
+from ddt import data
+from ddt import unpack
```
Я бы советовал ипортировать ddt.
И в последующем обращаться ```@ddt.data```, ```@ddt.unpack``` и т.д.
Это позволит повысить читаемость кода.

egor-makhlaev replied:
Принял к сведению.

egor-makhlaev committed:
Update tests

Shurke commented on rss_reader/__init__.py:
```diff
@@ -0,0 +1 @@
+from .rss_reader import *
```
А для чего это сделано?

egor-makhlaev replied:
Это устранило ошибки при запуске Тревисом.

Shurke commented on rss_reader/rss_reader/__init__.py:
```diff
@@ -0,0 +1 @@
+from .rss_reader import *
```
И тут еще раз.

egor-makhlaev replied:
Ответил выше.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,143 @@
+import sys
+import json
+import logging
+import argparse
+from pathlib import Path
+from itertools import islice
+from urllib.error import URLError
+from urllib.error import HTTPError
+from urllib.request import urlopen
+import xml.etree.ElementTree as ET
```
Небольшая придирка.
Если откыть РЕР8, то можно увидеть следующий текст.
Импорты должны быть сгруппированы в следующем порядке:

- импорты из стандартной библиотеки
- импорты сторонних библиотек
- импорты модулей текущего проекта

Вставляйте пустую строку между каждой группой импортов.

Пожалуйста, следуйте РЕР8.

egor-makhlaev replied:
Это всё импорты из стандартной библиотеки, если мне не изменяет память. Также помню, что нужно в алфавитном порядке, но привычка импорта "ёлочкой" не хочет уходить.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,143 @@
+
+        response = urlopen(url)
+
+    except HTTPError as e:
+        logger.error(f"The server couldn't fulfill the request.\nError code: {e.code}")
+        sys.exit()
+    except URLError as e:
+        logger.error(f"Failed to reach a server.\nReason: {e.reason}")
+        sys.exit()
```
А как вы обработаете ошибку, которая не относится к перечисленным?

egor-makhlaev replied:
Принял к сведению. Добавлю обработку общего случая.

Shurke commented on rss_reader/tests/test_main.py:
```diff
@@ -0,0 +1,37 @@
+from rss_reader import VERSION
+
+
+class TestMain(unittest.TestCase):
```
По какой причине нет позитивного теста?
Т.е. когда передали URL, вернули данные в нужном виде.

egor-makhlaev replied:
Принял к сведению. Добавлю позитивный тест. Я правильно понимаю, что это будет уже не unit, а integration тест, поскольку будет тестироваться взаимодействие нескольких функций?

Shurke replied:
Ну почему же. Можно просто использовать mock.

Shurke commented:
Если передать в вашу реализацию ридера пустую стороку, то он падает.
![Egor Makhlaev](https://user-images.githubusercontent.com/44495358/117570427-b61e6c80-b0d2-11eb-89d8-e16ab74bd41d.png)

Shurke commented:
Я думал, что при применении --json должен как-то измениться вывод. Или я не прав?
![Egor Makhlaev json](https://user-images.githubusercontent.com/44495358/117570445-d2baa480-b0d2-11eb-9112-25b6b1ab113d.png)

Shurke commented:
Наверное все же стоит как-то обрабатывать неправильные лимиты и сообщать пользователю о том, что он что-то не то ввел.
![Egor Makhlaev limit](https://user-images.githubusercontent.com/44495358/117570483-efef7300-b0d2-11eb-9a4a-47640e214655.png)

Shurke commented:
И завершающим комменратрием хочу обратить ваше внимание на следующее.
На лекции, посявщенный Git, мы рассматривали Gitflow Workflowl. Где говорилось о том, что в ветке main находится рабочая версия приложения. 
У вас это не выполняется, что не очень удобно для проверки.

egor-makhlaev committed:
Add functionality in LocalStorage class in local_storage.py

egor-makhlaev commented:
Добавлю сообщение при неправильно введённом лимите, а json сохраняется в директорию data как news.json, не счел нужным лишний раз "захламлять" stdout.

egor-makhlaev committed:
Add handling general exception in get_response of rss_reader.py

Shurke commented:
> Добавлю сообщение при неправильно введённом лимите, а json сохраняется в директорию data как news.json, не счел нужным лишний раз "захламлять" stdout.

В задании вполне ясное указание на этот счет
--json         Print result as JSON in stdout

egor-makhlaev commented:
> > Добавлю сообщение при неправильно введённом лимите, а json сохраняется в директорию data как news.json, не счел нужным лишний раз "захламлять" stdout.
> 
> В задании вполне ясное указание на этот счет
> --json Print result as JSON in stdout

Понял, добавлю печать json в stdout.

egor-makhlaev committed:
Add functionality in set_channel_by_url method of LocalStorage class in local_storage.py

egor-makhlaev committed:
Add warning of wrong limit in set_limit and print JSON to stdout in print_json

egor-makhlaev committed:
Fix build errors

egor-makhlaev committed:
Fix import errors

egor-makhlaev committed:
Fix import errors

egor-makhlaev committed:
Add news.json for tests

egor-makhlaev committed:
Fix news.json

egor-makhlaev committed:
Fix news.json

egor-makhlaev commented:
На модуль local_storage пока смотреть не нужно, это задел для 3 итерации.

egor-makhlaev committed:
1. Change the way that the news is represented
2. Implement printing and saving utf-8 texts with json.dumps as UTF8, not as \u escape sequence
3. Add functionality to LocalStorage class from local_storage module

egor-makhlaev committed:
Code refactoring in tests

Shurke commented:
А вот теперь при выполнение ```pip install .``` получается вот такая картина.
![Screenshot from 2021-05-11 18-39-27](https://user-images.githubusercontent.com/44495358/117844334-46ea7900-b288-11eb-8c04-ddc93e4b77df.png)

Shurke commented:
> На модуль local_storage пока смотреть не нужно, это задел для 3 итерации.

Вот для такого случая и нужно использовать ветки, а не работать в одну.
Иначе приложение ломается. И банально уже 1-2 итереацию недоправерить.
А так бы была в ветке main рабочая версия 1-2 итерации и параллельно шла бы работа над 3й в отдельной ветке.

egor-makhlaev commented:
> А вот теперь при выполнение `pip install .` получается вот такая картина.
> ![Screenshot from 2021-05-11 18-39-27](https://user-images.githubusercontent.com/44495358/117844334-46ea7900-b288-11eb-8c04-ddc93e4b77df.png)

Такой вывод получается если перед `pip install .` не вводить `pip install -r requirements.txt`, добавлю install_requires в setup.py. На счет ветки согласен, планирую добить 3 итерацию до рабочего состояния в этой, а 4 уже реализовывать в отдельной.

egor-makhlaev committed:
Update setup.py

Add install_requires argument

egor-makhlaev committed:
Create logger_config module which provides unique logger

Shurke commented:
> > А вот теперь при выполнение `pip install .` получается вот такая картина.
> > ![Screenshot from 2021-05-11 18-39-27](https://user-images.githubusercontent.com/44495358/117844334-46ea7900-b288-11eb-8c04-ddc93e4b77df.png)
> 
> Такой вывод получается если перед `pip install .` не вводить `pip install -r requirements.txt`, добавлю install_requires в setup.py. На счет ветки согласен, планирую добить 3 итерацию до рабочего состояния в этой, а 4 уже реализовывать в отдельной.

Справедливости ради стоит сказать, что изначально ваш rss_reader ставился именно `pip install .`. И работало все корректно.
Кроме того, это усложняет процесс установки:

1. сначала надо в папке PythonFinalTask выполнить `pip install -r requirements.txt`.
2. перейти в rss_reader и выполнить `pip install -r requirements.txt`.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,180 @@
+    channel_title = root.findtext("channel/title")
+    channel_items = []
+
+    for index_of_news_item, news_item in enumerate(root.iterfind("channel/item")):
+        logger.info(f"Process item № {index_of_news_item + 1}")
+
+        channel_items.append({
+            "Title": news_item.findtext("title"),
+            "Date": news_item.findtext("pubDate"),
+            "Link": news_item.findtext("link"),
+        })
+
```
Хочу предложить подумать над следующим моментом. 
Вроде как установил лимит 1, но в тоже время в логах видно, что процессятся все 50 новостей. Чисто исходя из логики, то не совсем понятно, зачем это делается.
![Screenshot from 2021-05-12 19-34-23](https://user-images.githubusercontent.com/44495358/118011898-2f2ef580-b359-11eb-9481-c73fa37ad32e.png)

egor-makhlaev replied:
> Хочу предложить подумать над следующим моментом.
> Вроде как установил лимит 1, но в тоже время в логах видно, что процессятся все 50 новостей. Чисто исходя из логики, то не совсем понятно, зачем это делается.
> ![Screenshot from 2021-05-12 19-34-23](https://user-images.githubusercontent.com/44495358/118011898-2f2ef580-b359-11eb-9481-c73fa37ad32e.png)

В моей реализации ридера --limit влияет только на печать новостей в stdout и вывод новостей в формате JSON в stdout, а обработка происходит для всех новостей, что пришли в response. Это сделано для того, чтобы в local storage попадало как можно больше новостей, чтобы впоследствии, при выборе по определенной дате, выборка была больше. Или же нужно кешировать только те новости, что были выведены в sdtout?

Shurke replied:
> В моей реализации ридера --limit влияет только на печать новостей в stdout и вывод новостей в формате JSON в stdout, а обработка происходит для всех новостей, что пришли в response. Это сделано для того, чтобы в local storage попадало как можно больше новостей, чтобы впоследствии, при выборе по определенной дате, выборка была больше. Или же нужно кешировать только те новости, что были выведены в sdtout?

Нет, ваша изначальная идея верна, что в кеш нужно складывать все, что вернулось из response.
Я бы сказал, что это больше придирка, т.к. новости легковесные и процессятся быстро.
С другой стороны, вводя лимит = 1, как-то сбивает столку что отдается столько информации.

egor-makhlaev replied:
> > В моей реализации ридера --limit влияет только на печать новостей в stdout и вывод новостей в формате JSON в stdout, а обработка происходит для всех новостей, что пришли в response. Это сделано для того, чтобы в local storage попадало как можно больше новостей, чтобы впоследствии, при выборе по определенной дате, выборка была больше. Или же нужно кешировать только те новости, что были выведены в sdtout?
> 
> Нет, ваша изначальная идея верна, что в кеш нужно складывать все, что вернулось из response.
> Я бы сказал, что это больше придирка, т.к. новости легковесные и процессятся быстро.
> С другой стороны, вводя лимит = 1, как-то сбивает столку что отдается столько информации.

Значит буду указывать конечное число обработанных новостей, к примеру: "Parsed 35 items from response"

egor-makhlaev committed:
Add docstrings to methods of LocalStorage class from local_storage module

egor-makhlaev committed:
1. Structure tests
2. Add test that tests get_response function from rss_reader module with bad url

egor-makhlaev committed:
Add logging to methods in LocalStorage class of local_storage module

egor-makhlaev committed:
Add tests to get_news_items_by_url_and_date method of LocalStorage class from local_storage module

egor-makhlaev committed:
Add ddt to test_get_response method of TestExceptions class from test_rss_reader module

egor-makhlaev committed:
Add description of local storage to README.md

egor-makhlaev commented:
Implemented iteration 3

Shurke commented:
> > А вот теперь при выполнение `pip install .` получается вот такая картина.
> > ![Screenshot from 2021-05-11 18-39-27](https://user-images.githubusercontent.com/44495358/117844334-46ea7900-b288-11eb-8c04-ddc93e4b77df.png)
> 
> Такой вывод получается если перед `pip install .` не вводить `pip install -r requirements.txt`, добавлю install_requires в setup.py.

Не исправлено.

Shurke commented:
Ридер несколько не рабочий.
![Screenshot from 2021-05-15 12-26-17](https://user-images.githubusercontent.com/44495358/118355461-c920b900-b578-11eb-86d6-58a49ad4c22b.png)

egor-makhlaev committed:
Merge branch 'final-task-implementation' of https://github.com/egor-makhlaev/PythonFinalTask into final-task-implementation

1. Update VERSION to 3.0
2. Fix UnboundLocalError in set_news_items_by_url method of LocalStorage class from local_storage module
3. Add pyproject.toml to fix issues with pip install .

egor-makhlaev committed:
Merge branch 'main' into final-task-implementation

egor-makhlaev committed:
Fix SyntaxError in setup.py

egor-makhlaev committed:
Correct logging in set_news_items_by_url method of LocalStorage class from local_storage module

egor-makhlaev committed:
Fix style issue in local_storage module

Shurke commented:
Мне кажется что команда и результат как-то между собой не совпадают.

UPD.
Так, я разобрался.
1. Дату все же можно в более читаемый формат вывести.
2. И стоит скорее добавить новое сообщение для логера, а не использовать от `--limit`. Что-то в духе "Найдено на такую дату ..." или "отфильтровано и получено столько-то записей".
![Screenshot from 2021-05-17 18-32-36](https://user-images.githubusercontent.com/44495358/118515815-65e08380-b73e-11eb-9290-e9b31dca2fb8.png)

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,149 @@
+try:
+    from logger_config import get_logger
+except ImportError:
+    from .logger_config import get_logger
```
Это кусок кода выглядит очень странно.

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,149 @@
+                            news_item {"Feed": (str), "Title", (str), "Date": (srt), "Link": (str)}: News item dict
+
+                    Returns:
+                            (datetime.datetime): Date from news item dict by key "Date"
```
А для чего двойные отступы?

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,149 @@
+
+            while storage_has_news_items_by_url and i < len(sorted_news_items) and \
+                    date_of_latest_news_item_by_url < self.parse_date_from_news_item(sorted_news_items[i]):
+                i += 1
```
Вот этот код прямо больно читать.

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,149 @@
+
+        def fold_news_items_by_url_and_date(news_items):
+            """Folds news items get by specific date to list named "news_items_by_specific_date"."""
+            nonlocal news_items_by_specific_date
```
Я не вижу смысла использовать здесь `nonlocal`.
Как бы можно просто вернуть список с результатами через `return`.
Кроме того, список является изменяемым объектом. Следовательно, просто передав его во внутрь, то он будет изменен нужным образом.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,178 @@
+    from logger_config import get_logger
+except ImportError:
+    from .local_storage import LocalStorage
+    from .logger_config import get_logger
```
Тоже не понятно, что и зачем.

Shurke commented on rss_reader/tests/test_rss_reader.py:
```diff
@@ -0,0 +1,127 @@
+from rss_reader.rss_reader import print_news
+from rss_reader.rss_reader import get_response
+from rss_reader.rss_reader import parse_response
+from rss_reader.rss_reader import limit_news_items
```
Вот эта часть сворачивается до одной строки.
```python
from rss_reader import  rss_reader
```
и дальше просто обращаться 
` rss_reader.main`

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,149 @@
+            """Folds news items get by specific date to list named "news_items_by_specific_date"."""
+            nonlocal news_items_by_specific_date
+
+            tmp = list(filter(lambda item: self.parse_date_from_news_item(item).date() == pub_date.date(), news_items))
```
Стоит отрефакторить данную строку в что-то более удобочитаемое.

Shurke commented on rss_reader/tests/test_local_storage.py:
```diff
@@ -0,0 +1,26 @@
+
+
+@ddt.ddt
+class TestGetNewsItemsByUrlAndDate(unittest.TestCase):
```
Относительно функционала, связанного с использованием локального хранилища, тестов слишком мало.

egor-makhlaev committed:
Create format conversion

egor-makhlaev committed:
1. Add required packages in requirements.txt and setup.py
2. Fix typo and style issues in setup.py

egor-makhlaev committed:
Move data directory inside rss_reader package

egor-makhlaev committed:
Use glob to select all subfolders in setup.py

egor-makhlaev committed:
Fix path to test data

egor-makhlaev committed:
1. Add get_news_items_images static method to LocalStorage class from local_storage module
2. Split data directory to project_data and package_data
3. Add image_url attribute to news_item in parse_response function from rss_reader module

egor-makhlaev committed:
Add docstrings and some code refactoring

egor-makhlaev committed:
1. Fix style issue in local_storage module
2. Add .gitkeep to images directory

egor-makhlaev committed:
Replace relative imports to absolute

egor-makhlaev committed:
Refactoring of imports in test_rss_reader module

egor-makhlaev committed:
1. Change return type of get_response function from rss_reader module
2. Change argument type of parse_response function from rss_reader module
3. Make .gitignore ignore images and localstorage

egor-makhlaev committed:
1. Add tests in test_local_storage module
2. Replace VERSION constant to rss_reader module
3. Refactor some code

egor-makhlaev committed:
Fix style issue in test_local_storage module and hardcore version argument in setup module

egor-makhlaev committed:
1. Refactor code in local_storage module
2. Add test for testing set_news_items_by_url method of LocalStorage
   class from local_storage module
3. Add url with empty list of news items in teststorage.json

egor-makhlaev committed:
Add logging in format_converter module

egor-makhlaev committed:
Add recursive creation of directories in create_file method of Converter class from format_converter module

egor-makhlaev committed:
Fix style issue in format_conver module

egor-makhlaev committed:
1. Fix NameError in format_converter module
2. Add description of command line arguments and JSON structure to README.md

Shurke commented:
Несколько замечаний сразу:

1. в help стоит добавить подробную инфорамцию о том, в каком формате нужно задавать путь для конвертации в файл.
2. стоит предусмотреть какое-то действие на то, если не был задан путь к папке, куда нужно сохранить файл.
3. перебрал больше десятка различных вариантов пути, ни один из нех не сработал.

![Screenshot from 2021-05-24 19-42-22](https://user-images.githubusercontent.com/44495358/119379766-8cfffd80-bcc8-11eb-9842-818413226f85.png)
![Screenshot from 2021-05-24 19-43-33](https://user-images.githubusercontent.com/44495358/119379769-8e312a80-bcc8-11eb-8645-b67adf4b5f52.png)
![Screenshot from 2021-05-24 19-44-55](https://user-images.githubusercontent.com/44495358/119379773-8e312a80-bcc8-11eb-85aa-d740f5ff7a7d.png)

egor-makhlaev commented:
> Несколько замечаний сразу:
> 
> 1. в help стоит добавить подробную инфорамцию о том, в каком формате нужно задавать путь для конвертации в файл.
> 2. стоит предусмотреть какое-то действие на то, если не был задан путь к папке, куда нужно сохранить файл.
> 3. перебрал больше десятка различных вариантов пути, ни один из нех не сработал.
> 
> ![Screenshot from 2021-05-24 19-42-22](https://user-images.githubusercontent.com/44495358/119379766-8cfffd80-bcc8-11eb-9842-818413226f85.png)
> ![Screenshot from 2021-05-24 19-43-33](https://user-images.githubusercontent.com/44495358/119379769-8e312a80-bcc8-11eb-8645-b67adf4b5f52.png)
> ![Screenshot from 2021-05-24 19-44-55](https://user-images.githubusercontent.com/44495358/119379773-8e312a80-bcc8-11eb-85aa-d740f5ff7a7d.png)

Вы указали аргументы `to-html` и `to-pdf`, а ридер ожидает `--to-html` и `--to-pdf`. Добавлю в описание ридера, что введёный путь будет приводиться к абсолютному.
![image](https://user-images.githubusercontent.com/68485879/119397201-7664a100-bcde-11eb-81fb-425e8a786490.png)

Shurke commented:
> Вы указали аргументы `to-html` и `to-pdf`, а ридер ожидает `--to-html` и `--to-pdf`. Добавлю в описание ридера, что введёный путь будет приводиться к абсолютному.

Справедливое замечание с вашей стороны. Виноват, был не внимателен.

egor-makhlaev committed:
1. In test_rss_reader module split TestExceptions class into TestGetResponse
and TestParseResponse
2. Add Mocking to test_get_response method of TestGerResponse class from
test_rss_reader module, so tests no longer require Internet connection

Shurke commented:
> > Вы указали аргументы `to-html` и `to-pdf`, а ридер ожидает `--to-html` и `--to-pdf`. Добавлю в описание ридера, что введёный путь будет приводиться к абсолютному.
> 
> Справедливое замечание с вашей стороны. Виноват, был не внимателен.

Так, в описании пример с описанием пути для конвертации так и не появился.

Shurke commented:
Странный результат.
![Screenshot from 2021-05-25 21-22-05](https://user-images.githubusercontent.com/44495358/119548830-4b8e5180-bd9f-11eb-9e3c-646b5a21a728.png)

egor-makhlaev commented:
> Странный результат.
> ![Screenshot from 2021-05-25 21-22-05](https://user-images.githubusercontent.com/44495358/119548830-4b8e5180-bd9f-11eb-9e3c-646b5a21a728.png)

У вас команда `rss_reader` повторяется.

Shurke commented:
> > Странный результат.
> > ![Screenshot from 2021-05-25 21-22-05](https://user-images.githubusercontent.com/44495358/119548830-4b8e5180-bd9f-11eb-9e3c-646b5a21a728.png)
> 
> У вас команда `rss_reader` повторяется.

Справедливо.

Shurke commented:
Я думаю, что тут больше выведено, чем должно быть.
Точно такой же вывод получается, если запрашивать новости через `--date`.
![Screenshot from 2021-05-25 21-50-16](https://user-images.githubusercontent.com/44495358/119552327-3e736180-bda3-11eb-89f3-5022dca65e4c.png)

Shurke commented:
Из Task clarification (IV):
> You can choose yourself the way in which the news will be displayed, but the final text result should contain pictures and links, if they exist in the original article and if the format permits to store this type of data.

![Screenshot from 2021-05-25 21-49-56](https://user-images.githubusercontent.com/44495358/119552868-acb82400-bda3-11eb-8989-72c82b04e8fc.png)

egor-makhlaev commented:
> Из Task clarification (IV):
> 
> > You can choose yourself the way in which the news will be displayed, but the final text result should contain pictures and links, if they exist in the original article and if the format permits to store this type of data.
> 
> ![Screenshot from 2021-05-25 21-49-56](https://user-images.githubusercontent.com/44495358/119552868-acb82400-bda3-11eb-8989-72c82b04e8fc.png)

Очень странно. Сходу даже и не сказать в каком направлении искать неполадку. 
Может помочь, если вы подскажите мне url картинки указанный в html, а так же url, занесённый в local storage (local storage хранится по адресу `место куда pip ставит пакеты/rss_reader/package_data/json/localstorage.json`). Ещё может помочь указание флага `--verbose`. Понимаю, что баги в своём коде искать должен я сам, но этот мне кажется очень тяжело воспроизводимым.
![image](https://user-images.githubusercontent.com/68485879/119554944-f3a71900-bda5-11eb-9c05-06705c0ef778.png)
UPD.
Я разобрался. Всё дело было в том, что на виртуальной машине я добавил (и благополучно забыл об этом) изображение в директорию `rss_reader/package_data/images/`, чтобы сохранить её при установке пакета, а поскольку все изображения в этой директории я "игнорю" в `.gitignore`, добавление изображения в директорию осталось без внимания для гита. Интересная ситуация получилась однако.

egor-makhlaev committed:
Add initial image to rss_reader/package_data/images/ to keep directory "images" when installing package

Shurke commented on rss_reader/tests/test_rss_reader.py:
```diff
@@ -0,0 +1,116 @@
+
+    @ddt.file_data("../project_data/json/test_get_response.json")
+    @patch.object(rss_reader, "urlopen")
+    def test_get_response(self, mocked_function, string_with_exception_creation, exception_attributes, output_message):
```
Название теста не совсем отражает смысл происходящего в тесте.
И стоит добавить позитивный тест.

egor-makhlaev replied:
Название изменил, позитивный тест добавил.

Shurke commented on rss_reader/tests/test_local_storage.py:
```diff
@@ -0,0 +1,64 @@
+        self.assertEqual(self.local_storage.get_number_of_news_items_by_url(url), 2, "Wrong number")
+
+
+if __name__ == "__main__":
+    print(get_data())
```
Я думаю, что тут все таки должно быть
```python
if __name__ == "__main__":
    unittest.main()
```

egor-makhlaev replied:
Исправил.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,192 @@
+    channel_title = root.findtext("channel/title")
+    news_items = []
+
+    for news_item in root.iterfind("channel/item"):
+        # Finds images only in response from https://news.yahoo.com/rss/
+        image_element = news_item.find('{http://search.yahoo.com/mrss/}content')
```
А почему другие ленты обделены вниманием?

egor-makhlaev replied:
Просмотрев несколько ресурсов, пока что не выявил закономерности, которая позволила бы извлекать картинки новостных элементов из них всех.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,192 @@
+
+
+def main(argv=sys.argv):
+    parser = argparse.ArgumentParser(description="Pure Python command-line RSS reader.")
+    parser.add_argument("source", nargs="?" if "--date" in argv else None, type=str, help="RSS URL")
+    parser.add_argument("--version", action="version", version=f'"Version {VERSION}"', help="Print version info")
+    parser.add_argument("--json", action="store_true", help="Print result as JSON in stdout")
+    parser.add_argument("--verbose", action="store_true", help="Outputs verbose status messages")
+    parser.add_argument("--limit", type=int, help="Limit news topics if this parameter provided")
+    parser.add_argument("--date", type=str, help="Return news topics which were published in specific date")
+    parser.add_argument("--to-html", type=validate_filepath_arg, help="Save news in .html format by provided path")
+    parser.add_argument("--to-pdf", type=validate_filepath_arg, help="Save news in .pdf format by provided path")
```
По-хорошему, то это отдельная функция, где создается ArgumentParser.

egor-makhlaev replied:
Вынес создание парсера в модуль helper.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,192 @@
+        news_items = local_storage.set_news_items_by_url(args.source, news_items)
+
+    news_items = limit_news_items(news_items, args.limit)
+    print_news(news_items)
+
+    if args.json:
+        print_json(news_items)
```
А для чего при передаче `--json` дважды выводить новости?
![Screenshot from 2021-05-26 09-04-07](https://user-images.githubusercontent.com/44495358/119610133-6ee9e880-be01-11eb-949e-3d5c72c9958b.png)

egor-makhlaev replied:
Согласен, дважды выводить смысла нет. Исправил.

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,179 @@
+        for news_item in news_items:
+            url = news_item["image_url"]
+
+            if url is not None:
```
Здесь можно просто `if url:` использовать.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,179 @@
+                storage_path_on_local_machine = get_path_to_data("images", url.split("/")[-1] + ".jpg")
+                news_item["image_url"] = storage_path_on_local_machine
+                executor.submit(urlretrieve, url, storage_path_on_local_machine)
+        # If wait in line below set to False, news is printed to stdout without waiting for all images to be downloaded
+        executor.shutdown(wait=True)  # If wait set to False, it can affect PDF generation (some images may be absent)
```
Более правильно будет оформить так
```python
# If wait in line below set to False, news is printed to stdout without waiting for all images to be downloaded
# If wait set to False, it can affect PDF generation (some images may be absent)
executor.shutdown(wait=True) 
```

egor-makhlaev replied:
Переоформил.

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,179 @@
+        # If wait in line below set to False, news is printed to stdout without waiting for all images to be downloaded
+        executor.shutdown(wait=True)  # If wait set to False, it can affect PDF generation (some images may be absent)
+
+    def __init__(self, name):
+        logger.info(f'Create local storage "{name}"')
+
+        storagepath = Path(get_path_to_data("json", f"{name}.json"))
+        storagepath.touch(exist_ok=True)
+
+        self.storagepath = storagepath
```
`__init__` стоит помещать сразу под объявлением класса.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,179 @@
+            max_index = None
+
+            if storage_has_news_items_by_url:
+
```
Здесь пустую строку можно убрать.

egor-makhlaev replied:
Принял к сведению.

Shurke commented on rss_reader/rss_reader/local_storage.py:
```diff
@@ -0,0 +1,179 @@
+        """
+        storage_content = self.read_from_storage_file()
+
+        if url in storage_content:
+            result = len(storage_content[url])
+        else:
+            result = 0
```
Такие вещи можно через тернарный оператор писать
```python
result = len(storage_content[url]) if url else 0
```

egor-makhlaev replied:
Принял к сведению, но нужно не забыть в if'е тернарного оператора убедиться, что url всё-таки присутствует в storage_content.
```python
result = len(storage_content[url]) if url in storage_content else 0
```

Shurke commented on rss_reader/rss_reader/format_converter.py:
```diff
@@ -0,0 +1,98 @@
+
+        return template.render(news_items=news_items)
+
+    def __init__(self, *, directory_path, file_name):
+        self.directory_path = Path(directory_path).absolute()
+        self.file_path = self.directory_path / file_name
```
`__init__` все же стоит писать сразу после `class Converter(abc.ABC):`

egor-makhlaev replied:
Принял к сведению.

egor-makhlaev committed:
1. Change method initialization order in LocalStorage and Converter
   classes
2. Add test_get_response_with_valid_url method to class from
   test_rss_reader module
3. Create function create_argument_parser in helper module that creates
   argument parser
4. Change the way of output JSON to stdout when --verbose argument is
   provided
5. Refactor some code
